rails_event_store_active_record/lib/rails_event_store_active_record/event_repository.rb

```diff
@@ -46,6 +46,17 @@ def read(specification)
       @repo_reader.read(specification)
     end
 
+    def update_messages(messages)
+      hashes = messages.map(&:to_h)
+      hashes.each{|h| h[:id] = h.delete(:event_id) }
+      for_update = messages.map(&:event_id)
+      start_transaction do
+        existing = Event.where(id: for_update).pluck(:id)
+        (for_update - existing).each{|id| raise RubyEventStore::EventNotFound.new(id) }
+        Event.import(hashes, on_duplicate_key_update: [:data, :metadata, :event_type])
```

Comment:
@paneq should this import in batches? How does this kind of rewriting of many events impact things?

Also, wonder if "rewrite" is a term to consider.

Reply:
@joelvh I don't understand this question "should this import in batches?"

Obviously I believed it should, otherwise I wouldn't write it that way. If you believe it shouldn't please state the reasons.

> How does this kind of rewriting of many events impact things?

What _things_ exactly do you have in mind ? How many is many? You can pass a single event here and you can pass a thousand as well. I was yesterday rewriting in batches of 10K and it worked fine for me.

> Also, wonder if "rewrite" is a term to consider.

sounds like a good option.

Reply:
@paneq right, my question wasn't clear. Looking at this method in isolation, the `start_transaction` part made me think I'd want to rewrite all events in a single transaction, so it made me think that this was handling all changes in one call. But what you're indicating makes sense, which is that you might batch on the read side, and call `update_messages` for each batch. If errors occur, then the transaction would roll back for that last batch.

Regarding "things", I was generalizing about what considerations should be called out in using this API around performance. Rewriting many many events in one transaction is probably something to be avoided. Rewriting in batches is more efficient and the code should handle old and new events until all events are successfully migrated.

Reply:
And regarding "many" -- that is obviously relative, point being whether some best practices should be called out (maybe in the docs) about how to use this API. My last comment clarifying what I meant with batches is related to clarifying how you would think about using this API for, let's say 100k-1m events. (I haven't used this volume myself yet, so I don't know what volume is a lot or an unreasonable amount of events to expect to process or load into memory.)

Hope that helps clarify the view point I was coming from.

Reply:
P.S. One last note: I haven't yet had to use all features of RES, particularly batching events and things related to retroactive projections. (I'm updating read models with event handlers only, right now.) I'm on the verge of diving into those features, so that is why maybe some of my questions sound naive in relation to RES and existing best practices. I am currently planning out building a system for rewriting events and also building up new read models from existing events (using RES projections) in the coming weeks. Maybe that will give me better context when I get into that.

Reply:
> Rewriting many many events in one transaction is probably something to be avoided. Rewriting in batches is more efficient and the code should handle old and new events until all events are successfully migrated.

Indeed. That's what's shown in docs. But it depends if you are doing live migration or have a planned downtime.

> How you would think about using this API for, let's say 100k-1m events

Tests on my local machine indicated that updating in batches of 1000 and 10_000 gave me the same performance. But 1000 was definitely faster than 100. I think you can apply all your intuitions and experiences from updating rows in any SQL db. 

The set:
* must fit into memory
* must not generate too big sql statement